grammars/parser.pegjs

```diff
@@ -1,121 +1,193 @@
-Program = _ e:Expression { return { type: "Root", expr: e } }
-
-Newline "newline"
-    = "\n"
-
-Space "space"
-    =  " "
-
-Whitespace = Newline / Space
-Anything = [^\n]*
-Comment "comment"
-    = "#" Anything Newline
-WS = Whitespace / Comment
-
-_ "whitespace" = WS*
-
-// Expression "expression"
-//     = BinExpr
-//     / OtherExpr
-
-Expression "expression"
-    = OtherExpr
-
-E = Expression
-
-BinExpr
-    = "<" _ "(" _ a:E b:E ")" _ { return {type: "Lt", a: a, b: b} }
-    / ">" _ "(" _ a:E b:E ")" _ { return {type: "Gt", a: a, b: b} }
-    / "*" _ "(" _ a:E b:E ")" _ { return {type: "Mul", a: a, b: b} }
-    / "/" _ "(" _ a:E b:E ")" _ { return {type: "Div", a: a, b: b} }
-    / "+" _ "(" _ a:E b:E ")" _ { return {type: "Add", a: a, b: b} }
-    / "-" _ "(" _ a:E b:E ")" _ { return {type: "Sub", a: a, b: b} }
-
-NoncallExpr = If / Let / Function / Literal / Identifier / ParenExpr
-OtherExpr = Call / NoncallExpr
-
-Argument "argument"
-    = e:Expression _ { return e }
+{
+    // Yikes... what a horrible import path.
+    var ast = require("../../../src/ast");
+
+    function foldLeft(f, z, xs) {
+        return xs.reduce(function(acc, x) {
+            return f(acc, x);
+        }, z);
+    }
 
-Call "function call"
-    = f:E "(" _ args:Argument* ")" _ {
-        return { type: "Call", f: f, args: args }
+    // Left-associative binary operator helper.
+    function lbo(a, xs) {
+        return foldLeft(function(acc, pair) {
+            return ast.BinOp(ast.Operator(pair[0]), acc, pair[1]);
+        }, a, xs);
+    }
+}
+
+Program
+    = _ "export" _ e:Expr _ { return ast.Module(e); }
+    / _ e:Expr _            { return ast.Script(e); }
+
+Keyword
+    = "if" / "then" / "else"
+    / "let" / "in"
+    / "true" / "false"
+    / "undefined" / "null"
+    / "export"
+
+Expr
+    = a:Expr0 xs:((o:";" _ b:Expr0) { return [o, b]; })*
+    { return lbo(a, xs); }
+
+Expr0
+    = "if"   _ p:Bop1
+      "then" _ t:Expr
+      "else" _ f:Expr
+    { return ast.If(p, t, f); }
+    / Expr1
+
+Expr1
+    = "let" _ "(" _ b:Bindings ")" _ "in" _ e:Expr
+    { return ast.Let(b, e); }
+    / Expr2
+
+Bindings
+    = b:Binding bs:(("," _ b2:Binding) { return b2; })*
+    { return [b].concat(bs); }
+
+Binding
+    = i:Identifier "=" _ e:Expr
+    { return ast.Binding(i, e); }
+
+Expr2
+    = Bop1
+
+b1 = "|>"
+b2 = "and" / "or"
+b3 = ">=" / "<=" / "<" / ">" / "=" / "!="
+b4 = "++"
+b5 = "+" / "-"
+b6 = "*" / "/"
+
+Bop1 = a:Bop2 xs:((o:b1 _ b:Bop2)  { return [o, b]; })* { return lbo(a, xs); }
+Bop2 = a:Bop3 xs:((o:b2 _ b:Bop3)  { return [o, b]; })* { return lbo(a, xs); }
+Bop3 = a:Bop4 xs:((o:b3 _ b:Bop4)  { return [o, b]; })* { return lbo(a, xs); }
+Bop4 = a:Bop5 xs:((o:b4 _ b:Bop5)  { return [o, b]; })* { return lbo(a, xs); }
+Bop5 = a:Bop6 xs:((o:b5 _ b:Bop6)  { return [o, b]; })* { return lbo(a, xs); }
+Bop6 = a:Bop7 xs:((o:b6 _ b:Bop7)  { return [o, b]; })* { return lbo(a, xs); }
```

Comment:
This part is *so* not ideal. Wish you could make a macro in PEG.js or something :\